Docs 7240 batch duke (#1207)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/batch-processing-concept.adoc

Mule allows you to process messages in batches.

> Within an application, you can initiate a Batch Job scope, which splits messages into individual records, performs actions upon each record, then reports on the results and potentially pushes the processed output to other systems or queues.

Reviewer: ```suggestion
Within an application, you can initiate a Batch Job scope, which splits messages into individual records, performs actions upon each record, and then reports on the results and potentially pushes the processed output to other systems or queues.
```

> Within an application, you can initiate a Batch Job scope, which is a block of code that splits messages into individual records, performs actions upon each record, then reports on the results and potentially pushes the processed output to other systems or queues.

Reviewer 2: remove `is a block of code that`

Author: done

Each batch step in a batch job contains processors that act upon a record to transform, route, enrich, or otherwise process data contained within it. By leveraging the functionality of existing Mule processors, the batch step offers a lot of flexibility regarding how a batch job processes records. See xref:batch-filters-and-batch-aggregator.adoc[Refining Batch Steps Processing] for more information.

A batch job executes when the flow reaches the process-records section of the batch job. When triggered, Mule creates a new batch job instance.
When the job instance becomes executable, the batch engine submits a task to the

> I/O pool to process each record. Parallelism occurs automatically, at the record

Reviewer 2: it's a task per each record block

Author: done

> block level. Auto-tuning logic controls the number or threads to use and the

Reviewer 2: Auto-tuning seems like a little out of the blue and out of context here. THis should link to the auto tuning documentation from the runtitme

Author: revised

Each batch step in a batch job contains processors that act upon a record to transform, route, enrich, or otherwise process data contained within it. By leveraging the functionality of existing Mule processors, the batch step offers a lot of flexibility regarding how a batch job processes records. See xref:batch-filters-and-batch-aggregator.adoc[Refining Batch Steps Processing] for more information.

A batch job executes when the flow reaches the process-records section of the batch job. When triggered, Mule creates a new batch job instance.
When the job instance becomes executable, the batch engine submits a task for each record block to the
I/O pool to process each record. Parallelism occurs automatically, at the record
block level. To control the number or threads to use and the rate at which these processing tasks execute,

> the Mule runtime engine automatically makes tuning decisions based on a set of processing types that are also configurable

Reviewer 2: `To control the number or threads to use and the rate at which these processing tasks execute,
the Mule runtime engine automatically makes tuning decisions based on a set of processing types that are also configurable` -> `The Mule Runtime uses its auto-tuning capabilities to determine how many threads are to be used and the level of parallelism to be applied`

Author: done

Each batch step in a batch job contains processors that act upon a record to transform, route, enrich, or otherwise process data contained within it. By leveraging the functionality of existing Mule processors, the batch step offers a lot of flexibility regarding how a batch job processes records. See xref:batch-filters-and-batch-aggregator.adoc[Refining Batch Steps Processing] for more information.

A batch job executes when the flow reaches the process-records section of the batch job. When triggered, Mule creates a new batch job instance.

> When the job instance becomes executable, the batch engine submits a task for each record block to the

Reviewer: Just curious: does a job have more than one instance?

Author: Yes, it does/can/certainly would in batch processing tons of record sets.

> I/O pool to process each record. Parallelism occurs automatically, at the record

Reviewer: Will your audience know what is meant by "parallelism occurs"? (cuz' I don't....)
I don't think of parallelism as being an action, just a result of some other action.

Reviewer: "Parallelism is applied automatically..."?

Author: They should understand. It's not really "applied." Things are processed in parallel instead of sequentially.

> block level. The Mule runtime engine uses its auto-tuning capabilities to determine how many threads to use and the level of parallelism to apply (see xref:execution-engine.adoc[Execution Engine]).  

Reviewer: We nearly never hyphenate after "auto":
```suggestion
block level. The Mule runtime engine uses its autotuning capabilities to determine how many threads to use and the level of parallelism to apply (see xref:execution-engine.adoc[Execution Engine]).  
```

In the On Complete phase, none of these variables (not event the original ones) are visible. Only the final result is available in this phase. Moreover, since the Batch Job Instance executes asynchronously from the rest of the flow, no variable set in either a Batch Step or the On Complete phase will be visible outside the Batch Scope.

=== Scheduling Strategy

A Mule application can contain several batch job definitions. Each definition 

> has its own scheduling strategy. A *Scheduling Strategy* enables you to control 

Reviewer: Why are we boldfacing and init. capping the second instance of "scheduling strategy"?
This use of the term is general, what the thing is, not specific to a UI label.

Author: one is a general use of the term, the other for the component in the UI. I know it looks weird but follows the guidelines, I think...

how instances of a given batch job are executed: 

> * ORDERED_SQUENTIAL (the default): If several job instances are in an executable

Reviewer: GLOBALLY convert all lists of this type, going forward, to appear as follows:
```suggestion
* ORDERED_SQUENTIAL (the default)
If several job instances are in an executable
```

Author: doing that requires a `+` after the bullet line to keep the formatting correct.

state at the same time, the instances execute one at a time based on their
creation timestamp.

> * ROUND_ROBIN: This setting attempts to execute all available instances of a

Reviewer: ditto
```suggestion
* ROUND_ROBIN
This setting attempts to execute all available instances of a
```

Author: same issue re formatting.

batch job using a round-robin algorithm to assign the available resources. 
+

> This option is useful when you can guarantee that no job execution can have a 

Reviewer: ```suggestion
The ROUND_ROBIN option is useful when you can guarantee that no job execution can have a 
```

> side effect on another job execution. Therefore, this option is not a good 

Reviewer: ```suggestion
affect another job execution. Therefore, this option is not a good 
```

Author: "side effect" is important here. It's widely understood.

Author: https://en.wikipedia.org/wiki/Side_effect_%28computer_science%29

choice for data synchronization jobs, which can update the same record in two 
concurrent jobs. Because the order of the job execution is not guaranteed, 

> you might end up with a prior version of the data. However, you can safely 

Reviewer: Because I believe that "end up with" is a bit idiomatic...:
```suggestion
your result might be a prior version of the data. However, you can safely 
```

In the On Complete phase, none of these variables (not event the original ones) are visible. Only the final result is available in this phase. Moreover, since the Batch Job Instance executes asynchronously from the rest of the flow, no variable set in either a Batch Step or the On Complete phase will be visible outside the Batch Scope.

=== Scheduling Strategy

The Batch Job scope provides Scheduling Strategy options for controlling the
execution of job instances.

* ORDERED_SQUENTIAL (the default): If several job instances are in an executable
state at the same time, the instances execute one at a time based on their
creation timestamp.
+

> Note that there is no guarantee that the records will be executed in order.

Reviewer 2: This doesn't belong here. This is true for all scheduling strategies.

Author: moved

This setting does not control the execution order of records in the batch job,
nor does it take into account the number of records each instance contains.
+
* ROUND_ROBIN: This setting attempts to execute all available instances of a
batch job using a round-robin algorithm to assign the available resources. It

> works for the use case in which jobs are triggered manually and you want them to

Reviewer 2: `It works for the use case in which jobs are triggered manually and you want them to
run concurrently,`. -> This is wrong. It should say:

`This option is useful when you can guarantee that no job execution can have a side effect on another one. For example, this is not a good fit for data synchronization jobs. This is because the same record could have been updated in two concurrent jobs, and since their ordering is not guarantee, you might end up with a prior version of the data. However, if your batch job is just picking up new records from a database or lifting individual files from an SFTP server (and you're sure that all records are completely independent), then you can safely use this strategy to parallelize the job's execution`

Author: ok, cool. Pls note that the blog says this, "In the December 2013 release there was no way to configure that. By default, the batch module would try to execute all available instances by using a round robin algorithm to assign the available resources. This is fine for the use case in which jobs are triggered manually and you want them to run concurrently. But there’re other uses cases, in which you need the first job instance to finish before allowing any others to run. An example of this situation is a one-way sync between two systems." and "Round Robin: This is the behaviour we used to have in December 2013 release."

Author: done

run concurrently, such as one-way synchronization. This setting was introduced
before the current default, ORDERED_SQUENTIAL.

> A Mule application can contain more than one batch job configuration. These

Reviewer 2: `batch job configuration` -> `batch job`

Author: done

> separate batch jobs can run at the same time because the scope of each

Reviewer 2: This entire paragraph is incorrect, what's said here is not true. It should just say that each app can contain several batch job definitions and that each one of them has its own scheduling strategy.

Author: done

In the On Complete phase, none of these variables (not event the original ones) are visible. Only the final result is available in this phase. Moreover, since the Batch Job Instance executes asynchronously from the rest of the flow, no variable set in either a Batch Step or the On Complete phase will be visible outside the Batch Scope.

=== Scheduling Strategy

A Mule application can contain several batch job definitions. Each definition 

> has its own scheduling strategy. The Batch Job scope provides 

Reviewer 2: `The Batch Job scope provides  *Scheduling Strategy* options for controlling the execution of job instances: ` -> `Scheduling Strategy allows controlling how instances of a given Batch Job are executed`

Author: done

*Scheduling Strategy* options for controlling the execution of job instances: 

* ORDERED_SQUENTIAL (the default): If several job instances are in an executable
state at the same time, the instances execute one at a time based on their
creation timestamp.
* ROUND_ROBIN: This setting attempts to execute all available instances of a
batch job using a round-robin algorithm to assign the available resources. 
+
This option is useful when you can guarantee that no job execution can have a 
side effect on another job execution. Therefore, this option is not a good 
choice for data synchronization jobs, which can update the same record in two 
concurrent jobs. Because the order of the job execution is not guaranteed, 
you might end up with a prior version of the data. However, you can safely 
use this strategy to parallelize the job's execution if your batch job 
retrieves only new records from a database or lifts individual files 
from an SFTP server (and you are certain that all records are completely 
independent).
+

> This setting was introduced before the current default, ORDERED_SQUENTIAL.

Reviewer 2: remove this line

Author: done

> Note that there is no guarantee that records will be executed in order.

Reviewer 2: `Note that there is no guarantee ` -> `Note that none of these strategies guarantee`

Author: done

The strategies do not control the execution order of records in the batch job,

> nor do they take into account the number of records each instance contains.

Reviewer 2: `nor do they take into account the number of records each instance contains.` -> `nor do they depend on the  number of records each instance contains.`

Author: done
